src/protocol/attribution/accumulate.rs

```diff
@@ -0,0 +1,322 @@
+        //
+        // Each list element interacts with exactly one other element in each iteration, and the interaction do not
+        // depend on the calculation results other elements, allowing the algorithm to be executed in parallel.
+
+        // generate powers of 2 that fit into input len. If num_rows is 15, this will produce [1, 2, 4, 8]
+        for step_size in std::iter::successors(Some(1u32), |prev| prev.checked_mul(2))
+            .take_while(|&v| v < num_rows)
+        {
+            let end = num_rows - step_size;
+            let mut accumulation_futures = Vec::with_capacity(end as usize);
+
+            // for each input row, create a future to execute secure multiplications
+            for i in 0..end {
+                let current = AccumulationInputRow {
+                    stop_bit: stop_bits[i],
+                    credit: credits[i],
+                    report: self.input[i],
+                };
+                let successor = AccumulationInputRow {
+                    stop_bit: stop_bits[i + step_size],
+                    credit: credits[i + step_size],
+                    report: self.input[i + step_size],
+                };
+
+                accumulation_futures.push(self.get_accumulated_credit(
+                    ctx,
+                    step_size,
+                    current,
+                    successor,
+                    RecordId::from(i),
+                    [steps[1], steps[2], steps[3], steps[4]],
+                ));
+            }
+
+            let results = try_join_all(accumulation_futures).await?;
+
+            // save the calculation results to these support vectors for use in the next iteration
+            results
+                .into_iter()
+                .enumerate()
+                .for_each(|(i, (credit, stop_bit))| {
+                    credits[i] += credit;
+                    stop_bits[i] = stop_bit;
+                });
+        }
+
+        // drop irrelevant fields and add another supporting field called `aggregation_bit` for the next step
+        let output: Batch<AccumulationOutputRow<F>> = self
+            .input
+            .iter()
+            .enumerate()
+            .map(|(i, x)| AccumulationOutputRow {
+                breakdown_key: x.breakdown_key,
+                credit: credits[i],
+                aggregation_bit: one,
+            })
+            .collect::<Vec<_>>()
+            .into();
+
+        // TODO: Append unique breakdown_key values at the end of the output vector for the next step
+        // Since we cannot see the actual breakdown key values, we'll append shares of [0..MAX]. Adding u8::MAX
+        // number of elements to the output won't be a problem. Adding u32::MAX elements will be 1B + u32::MAX, which
+        // exceeds our current assumption of `input.len() < 1B`.
+
+        Ok(output)
+    }
+
+    async fn get_accumulated_credit<S, N>(
+        &self,
+        ctx: &ProtocolContext<'_, S, N>,
+        step_size: u32,
```

Comment:
Rather than pass in the step size (which it doesn't really need to know), perhaps just pass in a boolean indicating if this is the first step or not.

Reply:
Yes, I was wondering what would be the most non-absurd way to pass this. I thought `if step_size > 1` was a tiny bit more readable than `if !first_iteration`, but I'm okay with a boolean, too.

Reply:
You could do something similar to what @richajaindce did here: https://github.com/martinthomson/raw-ipa/pull/97/files#diff-894d6d54e26cbe3155f95c21f6bd6b557778f2705f8371fc2bc1d31b2292cb74

She made functions accept an enum - which only had two values. Direction::LEFT and Direction::RIGHT.

This makes reading the call site easier to understand than seeing the function accept "true" or "false" for the nth argument. You could have FooBar::FIRST_ITERATION and FooBar::SUBSEQUENT_ITERATION or something.

Reply:
Hmm, that might be an overkill for this case. Rust doesn't support ternary conditional operator so the call site would then look like this.

```
let iteration = if step_size > 1 { FooBar::FIRST_ITERATION } else { FooBar::SUBSEQUENT_ITERATION }; // need the brackets, too

get_accumulated_credit(
    ctx,
    iteration,
    ...
```

Reply:
You could have a mutable local variable initialized to FooBar::FIRST_ITERATION, which is updated to FooBar::SUBSEQUENT_ITERATION at the end of the loop. Then you don't need a ternary.

Reply:
Or maybe I could use the `steps` to infer which iteration I'm in? As I mentioned in a separate comment, I think I need 4 * log(n) steps rather than just 4. I need to look into Martin's PR to see if that's feasible.